WIP: Copy license files

matthiasbeyer opened the pull request:
**Note: ** This is based on #198 and #199, so those should be merged first. **Do not merge**

This is my first attempt on replacing the `license_file` logic with the `copy_dir()` function which honors include and exclude globs.

## Decision needed

For the following points, decisions are needed:

* The actual syntax of the "not" operator. I used tilde (`~`) for now, this should of course move to a global constant and not be inlined
* The `copy_dir()` function does not report missing files on the "source" side. The previous implementation of `copy_license_files()` returned an error when the license file to be copied was not found in the source directory(/directories). That part is gone with this change. We probably want to keep that bit though.

---

Overall this is **WIP** and most likely does not yet work. I wanted to PR this so you can have a look whether the direction is good or not.

wolfv commented:
- regarding the warning: if we can figure out that a single file is pointed to (not ending with `/` or `*` then we could still warn if the file does not exist (e.g. if you have `- LICENSE`, and `LICENSE` does not exist, pls warn).
- we could also warn if a positive glob does not match any files at all
- Inside the copy-licenses function we shoudl also warn if `0` licenses are selected

Looks good! I am happy to see progress. Some tests might be good, too.

matthiasbeyer commented:
Updated the PR. Sorry for the delay, I was traveling (Brussels, Eurorust).
Please have a look, I implemented the three points you mentioned.

One thing has to be mentioned though:

> * we could also warn if a positive glob does not match any files at all

This was unnecessarily hard to get implemented. Problem is that the `WalkBuilder::filter_entry()` predicate must be a pure function, so we cannot pass in a bool flag to check whether any "include glob" has matched. The glob lists have to be moved into that predicate, which makes it inaccessible afterwards, so we cannot check later - only if we _clone_ the include glob list and use the cloned list.
That is what I did: I put the include glob list behind an `Arc` and used a clone of that to check later in the iterator code whether an include glob matched any file. This is not nice code, but I currently do not see any way to improve this.

wolfv commented:
Awesome, hope you enjoyed EuroRust. We really liked Ashley's talk about "it's all in the delivery" :)

Looks good to me – I'll try this a bit to see if it works as I am expecting and then will merge. I can run `cargo fmt` to fix the build issue.

matthiasbeyer commented:
Ah, always the formatter. I will do better next time :laughing: 

Yeah Eurorust was absolutely great!

wolfv commented:
i found a little issue with the code. I think `filter` prevents the WalkDir to recurse into directories. I was testing the following modified xtensor recipe:

```yaml
context:
  name: xtensor
  version: "0.24.6"

package:
  name: "{{ name|lower }}"
  version: "{{ version }}"

source:
  url: https://github.com/xtensor-stack/xtensor/archive/{{ version }}.tar.gz
  sha256: f87259b51aabafdd1183947747edfff4cff75d55375334f2e81cee6dc68ef655

build:
  number: 0
  script:
    - sel(unix):
        echo "boop"

requirements:

about:
  home: https://github.com/xtensor-stack/xtensor
  license: BSD-3-Clause
  license_file:
    - LICENSE
    - cmake/
    - docs/*.yml
  summary: The C++ tensor algebra library

```

wolfv commented:
I continued to make the function uglier by continuing to recurse into directories and stripping the `from` prefix from the paths (so that the glob matching works).
I also had to wrap the `from_copy` in an Arc<PathBuf>.

Let me know if you have any ideas :) I'll also discuss with @baszalmstra 

```
            if let Some(ft) = entry.file_type().as_ref() {
                if ft.is_dir() {
                    return true;
                }
            }

            println!("{}", entry.path().display());
            let stripped_path = entry.path().strip_prefix(from_copy.as_ref()).unwrap();
            (include_globs.len() == 0 || include_globs.is_match(stripped_path))
                && !exclude_globs.is_match(entry.path())
```

baszalmstra commented on src/packaging.rs:
```diff
@@ -511,37 +511,57 @@ fn copy_license_files(
+        {
+            let filepath = licenses_folder.join(license_glob);
+            if !filepath.exists() {
+                tracing::warn!(path = %filepath.display(), "File does not exist");
```
Shouldnt this result in an error?

baszalmstra replied:
Shouldve read the PR description better..

baszalmstra commented on src/packaging.rs:
```diff
@@ -511,37 +511,57 @@ fn copy_license_files(
+        .collect::<Vec<PathBuf>>();
+
+        if copied_files.is_empty() {
+            tracing::warn!("No license files were copied");
```
Shouldnt this also be an error? I assume it means that a user listed a license glob but it doesn't match anything. It would assume that would result in an error since its not what the user expects.

wolfv commented:
I have special cased folder globs now so that they work (means that we want to recursively include the entire folder).
I also take the relative path so that relative globs work.

LMK what you think @matthiasbeyer.

wolfv commented:
I merged this so that we can continue iterating on #205 – but we should still improve on this PR incl. the error messages. I'll open an issue for that.
